src/main/java/com/google/api/gax/retrying/RetryHandler.java

```diff
@@ -0,0 +1,101 @@
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.retrying;
+
+import com.google.api.gax.core.RetrySettings;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Future;
+
+/**
+ * A retry handler is responsible for the following operations:
+ * <ol>
+ * <li>Accepting or rejecting a task for retry, depending on the previous attempt result (exception)
+ * and/or other attempt properties (like number of already executed attempts or total time spent
+ * retrying).</li>
+ *
+ * <li>Creating first attempt {@link RetryFuture}, which acts as a facade, hiding from client code
+ * the actual scheduled retry task.</li>
+ *
+ * <li>Creating {@link RetryAttemptSettings} for each subsequnt retry attempt.</li>
+ *
+ * <li>Executing the actual {@link Callable} in a retriable context.</li>
+ *
+ * </ol>
+ *
+ * This interface is for internal/advanced use only.
+ *
+ * @param <ResponseT> response type
+ */
+public interface RetryHandler<ResponseT> {
+
+  /**
+   * Returns {@code true} if another attempt should be made, or {@code false} otherwise
+   *
+   * @param e exception thrown by the previous attempt
+   * @param nextAttemptSettings attempt settings, which will be used for the next attempt, if
+   * accepted
+   * @return {@code true} if another attempt should be made, or {@code false} otherwise
+   */
+  boolean accept(Throwable e, RetryAttemptSettings nextAttemptSettings);
+
+  /**
+   * Creates a first try {@link RetryFuture}, which is a facade, returned to the client code to wait
+   * for any retriable operation to complete.
+   *
+   * @param callable the actual callable, which should be executed in a retriable context
+   * @param globalSettings global retry settings (attempt independent)
+   * @return retriable future facade
+   */
+  RetryFuture<ResponseT> createFirstAttempt(
+      Callable<ResponseT> callable, RetrySettings globalSettings);
+
+  /**
+   * Creates next attempt {@link RetryAttemptSettings}, which defines properties of the next
+   * attempt. Eventually this object will be passed to
+   * {@link RetryHandler#accept(Throwable, RetryAttemptSettings)} and
+   * {@link RetryHandler#executeAttempt(Callable, RetryAttemptSettings)}.
+   *
+   * @param e exception thrown by the previous attempt
+   * @param prevSettings previous attempt settings
+   * @return next attempt settings
+   */
+  RetryAttemptSettings createNextAttemptSettings(Throwable e, RetryAttemptSettings prevSettings);
+
+  /**
+   * Executes an attempt. A typical implementation will either try to execute in the current thread
+   * or schedule it for an execution, using some sort of async execution service.
+   *
+   * @param callable the actual callable to execute
+   * @param attemptSettings current attempt settings
+   * @return the {@link Future}, representing the scheduled execution
+   */
+  Future<ResponseT> executeAttempt(
```

Comment:
Because it does not return ApiFuture. In typical case it returns whatever Future is returned by the underlining ShceduledExecutorService. I can wrap it in ApiFuture, but trying to avoid wrappers when they are not necessary (there are already wrappers around everything, including wrappers around wrappers).